File: python/0020有效的括号.py
s="}"

# 栈
def isValid(s: str) -> bool:
    m={'(':')','[':']','{':'}'}

    st=[]

    for c in s:
        if(c in m.keys()):
            st.append(c)
        elif(not st or m[st.pop()]!=c):
            return False
        
    return True if not st else False
            



print(isValid(s))

File: cpp/0020有效的括号.cpp
#include <iostream>
#include<stack>
#include<map>
using namespace std;


// 栈
bool isValid(string s)
{
    map<char, char> m = {
        {'[',']'},
        {'{','}'},
        {'(',')'}
    };

    stack<char> st;

    if(m.find(s[0])==m.end())
        return false;
    
    st.push(s[0]);

    for (int i = 1;i<s.size();++i)
    {
        if (!(m.find(s[i]) == m.end()))
            st.push(s[i]);
        
        else
        {
            if(st.empty()||m[st.top()]!=s[i])
                return false;

            st.pop();
        }
    }

    return st.empty() ? true : false;
}

int main()
{
    string s = "(){}}{";
    cout << isValid(s);
}